src/tools/dashboard/src/reference.rs

```diff
@@ -0,0 +1,280 @@
+        }
+    }
+    map
+}
+
+/// Extracts examples from the given relative `paths` in the `book_dir` and
+/// saves them in `gen_dir`.
+fn extract_examples(paths: Vec<&PathBuf>, book_dir: &Path, gen_dir: &Path) {
+    for path in paths {
+        let mut cmd = Command::new("rustdoc");
+        cmd.args([
+            "+nightly",
+            "--test",
+            "-Z",
+            "unstable-options",
+            book_dir.join(path).to_str().unwrap(),
+            "--test-builder",
+            &["src", "tools", "dashboard", "print.sh"]
+                .iter()
+                .collect::<PathBuf>()
+                .to_str()
+                .unwrap(),
+            "--persist-doctests",
+            gen_dir.to_str().unwrap(),
+            "--no-run",
+        ]);
+        cmd.stdout(Stdio::null());
+        cmd.spawn().unwrap().wait().unwrap();
+    }
+}
+
+/// Copies the extracted rust code in `from_dir` to `src/test` following the
+/// hierarchy specified by `map`.
+fn organize_examples(map: &HashMap<PathBuf, PathBuf>, book_dir: &Path, from_dir: &Path) {
+    // The names of the extracted examples generated by `rustdoc` have the
+    // format `<path>_<line-num>_<test-num>` where occurrences of '/', '-', and
+    // '.' in <path> are replaced by '_'. This transformation is not injective,
+    // so we cannot map those names back to the original markdown file path.
+    // Instead, we apply the same transformation on the keys of `map` in the for
+    // loop below and lookup <path> in those modified keys.
+    let mut modified_map = HashMap::new();
+    for (path, hierarchy) in map.iter() {
+        modified_map.insert(
+            book_dir.join(path).to_str().unwrap().replace(&['\\', '/', '-', '.'][..], "_"),
+            hierarchy.clone(),
+        );
+    }
+    for dir in from_dir.read_dir().unwrap() {
+        let dir = dir.unwrap().path();
+        // Some directories do not contain tests because the markdown file
+        // instructs `rustdoc` to ignore those tests.
+        if let Some(example) = dir.read_dir().unwrap().next() {
+            let example = example.unwrap().path();
+            copy(&example, &modified_map);
+        }
+    }
+}
+
+/// Copy the file specified by `from` to the corresponding location specified by
+/// `map`.
+fn copy(from: &Path, map: &HashMap<String, PathBuf>) {
+    // The path specified by `from` has the form:
+    // `src/tools/dashboard/target/ref/<key>_<line-num>_<test-num>/rust_out`
+    // We copy the file in this path to a new path of the form:
+    // `src/test/<val>/<line-num>.rs
+    // where `map[<key>] == <val>`. We omit <test-num> because all tests have
+    // the same number, 0.
+    // Extract `<key>_<line-num>_<test-num>`.
+    let key_line_test = from.parent().unwrap().file_name().unwrap().to_str().unwrap();
+    // Extract <key> and <line-num> from `key_line_test` to get <val> and
+    // construct destination path.
+    let _2 = key_line_test.rfind("_").unwrap();
```

Comment:
Could you use https://doc.rust-lang.org/std/string/struct.String.html#method.split ?

Reply:
Good idea! `<key>` contains 5+ underscores. So it's tricky to use this version. But I think, `rsplitn` will do the trick.